File: apps/watch/IronProtocolWatch/Models/Workout.swift
import Foundation

/// Hand-mirror of the workout-related schemas from
/// `packages/api-contract/src/schemas/workout.ts`.
///
/// Watch app Phase 1 only renders summary info (block name, day label,
/// readiness coefficient). Phase 2 will surface the full prescription list
/// and allow set logging from the wrist.

struct ExercisePrescription: Codable, Identifiable, Equatable {
    let id: String
    let sessionId: String
    let exerciseName: String
    let exerciseOrder: Int
    let prescribedSets: Int
    let prescribedReps: Int
    let prescribedRPE: Double
    let percentOfE1RM: Double?
    let targetWeightLbs: Double?
    let isAccessory: Bool
}

struct CompletedSet: Codable, Identifiable, Equatable {
    let id: String
    let sessionId: String
    let prescriptionId: String
    let setNumber: Int
    let reps: Int
    let weightLbs: Double
    let rpe: Double?
    let completedAt: String
}

struct TrainingSession: Codable, Identifiable, Equatable {
    let id: String
    let blockId: String
    let weekNumber: Int
    let dayNumber: Int
    let scheduledDate: String
    let completedAt: String?
    let readinessScore: Double?
    let readinessCoeff: Double
    let autoRegNote: String?
    let createdAt: String
    let prescriptions: [ExercisePrescription]
    let completedSets: [CompletedSet]
}

struct RegulatedPrescription: Codable, Equatable {
    let exerciseName: String
    let sets: Int
    let reps: Int
    let rpe: Double
    let percentOfE1RM: Double?
    let isAccessory: Bool
    let adjustedSets: Int
    let adjustedRpe: Double
    let adjustedPercentE1RM: Double?
    let targetWeightLbs: Double?
    let regulationNote: String
}

struct WorkoutBlockSummary: Codable, Equatable {
    let id: String
    let name: String
    let phase: String // HYPERTROPHY | STRENGTH | PEAKING | DELOAD
    let currentWeek: Int
    let currentDay: Int
    let weekCount: Int
}

/// Mirrors `WorkoutResponseSchema`.
struct WorkoutResponse: Codable, Equatable {
    let session: TrainingSession
    let label: String
    let readiness: Readiness?
    let regulated: [RegulatedPrescription]
    let block: WorkoutBlockSummary
}

File: packages/api-contract/src/schemas/workout.ts
import { z } from "zod";
import { IsoDateString } from "./common";
import { ReadinessSchema } from "./dashboard";

/**
 * GET /api/workout — today's auto-regulated session.
 * Source of truth: `apps/web/src/app/api/workout/route.ts` (GET).
 *
 * SwiftMirror: `IronProtocolWatch/Models/Workout.swift`
 */

export const ExercisePrescriptionSchema = z.object({
  id: z.string(),
  sessionId: z.string(),
  exerciseName: z.string(),
  exerciseOrder: z.number().int(),
  prescribedSets: z.number().int(),
  prescribedReps: z.number().int(),
  prescribedRPE: z.number(),
  percentOfE1RM: z.number().nullable(),
  targetWeightLbs: z.number().nullable(),
  isAccessory: z.boolean(),
});
export type ExercisePrescription = z.infer<typeof ExercisePrescriptionSchema>;

export const CompletedSetSchema = z.object({
  id: z.string(),
  sessionId: z.string(),
  prescriptionId: z.string(),
  setNumber: z.number().int(),
  reps: z.number().int(),
  weightLbs: z.number(),
  rpe: z.number().nullable(),
  completedAt: IsoDateString,
});
export type CompletedSet = z.infer<typeof CompletedSetSchema>;

export const TrainingSessionSchema = z.object({
  id: z.string(),
  blockId: z.string(),
  weekNumber: z.number().int(),
  dayNumber: z.number().int(),
  scheduledDate: IsoDateString,
  completedAt: IsoDateString.nullable(),
  readinessScore: z.number().nullable(),
  readinessCoeff: z.number(),
  autoRegNote: z.string().nullable(),
  createdAt: IsoDateString,
  prescriptions: z.array(ExercisePrescriptionSchema),
  completedSets: z.array(CompletedSetSchema),
});
export type TrainingSession = z.infer<typeof TrainingSessionSchema>;

export const RegulatedPrescriptionSchema = z.object({
  exerciseName: z.string(),
  sets: z.number().int(),
  reps: z.number().int(),
  rpe: z.number(),
  percentOfE1RM: z.number().nullable(),
  isAccessory: z.boolean(),
  adjustedSets: z.number().int(),
  adjustedRpe: z.number(),
  adjustedPercentE1RM: z.number().nullable(),
  targetWeightLbs: z.number().nullable(),
  regulationNote: z.string(),
});
export type RegulatedPrescription = z.infer<typeof RegulatedPrescriptionSchema>;

export const WorkoutBlockSummarySchema = z.object({
  id: z.string(),
  name: z.string(),
  // Phase enum: HYPERTROPHY | STRENGTH | PEAKING | DELOAD
  phase: z.string(),
  currentWeek: z.number().int(),
  currentDay: z.number().int(),
  weekCount: z.number().int(),
});
export type WorkoutBlockSummary = z.infer<typeof WorkoutBlockSummarySchema>;

export const WorkoutResponseSchema = z.object({
  session: TrainingSessionSchema,
  label: z.string(),
  readiness: ReadinessSchema.nullable(),
  regulated: z.array(RegulatedPrescriptionSchema),
  block: WorkoutBlockSummarySchema,
});
export type WorkoutResponse = z.infer<typeof WorkoutResponseSchema>;
